File: 12_Advanced_Graphs/06_Cheapest_Flights_Within_K_Stops/0787-cheapest-flights-within-k-stops.py
"""
Problem: LeetCode 787 - Cheapest Flights Within K Stops

Key Idea:
The problem can be solved using Dijkstra's algorithm with a modified priority queue (min-heap) and BFS (Breadth-First Search) approach. We can create a graph where each node represents a city, and the edges represent flights between cities with associated costs. We use a priority queue to explore the nodes in a way that prioritizes the minimum cost. We continue the BFS process until we reach the destination city or exhaust the maximum number of allowed stops (K+1).

Time Complexity:
- The priority queue operations take O(E * log(V)) time, where E is the number of flights and V is the number of cities.
- In the worst case, all cities are explored, and all flights are considered.
- Overall, the time complexity is O(E * log(V)).

Space Complexity:
- The space required for the graph representation is O(V + E), where V is the number of cities and E is the number of flights.
- The space for the priority queue is O(V).
- Other data structures used take constant space.
- Therefore, the space complexity is O(V + E).
"""

import heapq
import math
from typing import List


class Solution:
    def findCheapestPrice(
        self, n: int, flights: List[List[int]], src: int, dst: int, max_stops: int
    ) -> int:
        graph = [[] for _ in range(n)]
        min_heap = [
            (0, src, max_stops + 1)
        ]  # (total_cost, current_city, remaining_stops)
        distances = [[math.inf] * (max_stops + 2) for _ in range(n)]

        for u, v, w in flights:
            graph[u].append((v, w))

        while min_heap:
            total_cost, current_city, remaining_stops = heapq.heappop(min_heap)
            if current_city == dst:
                return total_cost
            if remaining_stops > 0:
                for neighbor, cost in graph[current_city]:
                    new_cost = total_cost + cost
                    if new_cost < distances[neighbor][remaining_stops - 1]:
                        distances[neighbor][remaining_stops - 1] = new_cost
                        heapq.heappush(
                            min_heap, (new_cost, neighbor, remaining_stops - 1)
                        )

        return -1

File: 12_Advanced_Graphs/06_Cheapest_Flights_Within_K_Stops/0787-cheapest-flights-within-k-stops.rs
/*
Problem: LeetCode 787 - Cheapest Flights Within K Stops

Key Idea:
The key idea is to use a modified form of Dijkstra's algorithm to find the cheapest price from the source to the destination while considering the maximum number of stops.

Approach:
1. Create an adjacency list representation of the graph where each node is an airport, and each edge is a flight with its cost. You can use a vector of vectors for this purpose. The index of the outer vector represents the airport, and each inner vector contains pairs (to, price) representing flights from that airport.
2. Initialize a vector stops to keep track of the minimum number of stops required to reach each airport from the source
3. Initialize all elements to i32::MAX initially, except for the source airport, which should be set to 0.
4. Initialize a priority queue (min-heap) min_heap to store tuples (total_price, airport, stops) in reverse order of total_price (i.e., lowest total_price comes first).
5. Push a tuple (0, src, 0) onto the min_heap, where 0 is the total price, src is the source airport, and 0 is the number of stops.
6. Explore Nodes in Priority Queue:
  - While the min_heap is not empty:
      - Pop the tuple (total_price, from, steps) from the min_heap in reverse order.
          - If steps exceeds k or if from is equal to dst, return total_price as the answer.
          - Otherwise, update stops[from] with steps if steps is smaller than the current value.
          - For each flight (to, price) from from to other airports:
              - Calculate the new_total_price as total_price + price.
              - Push the tuple (new_total_price, to, steps + 1) onto the min_heap.
7. If no valid path is found after exploring all nodes, return -1.

Time Complexity:
O(E + V * log(V)), where E is the number of flights (edges) and V is the number of airports (nodes). The use of a priority queue ensures that we explore nodes in order of increasing total_price.

Space Complexity:
O(V), for stops vector and the min_heap.
*/

use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        let mut graph = vec![vec![]; n as usize];
        for flight in &flights {
            let (from, to, price) = (flight[0] as usize, flight[1] as usize, flight[2]);
            graph[from].push((to, price));
        }

        let mut stops = vec![i32::MAX; n as usize];

        let mut min_heap = BinaryHeap::new();
        min_heap.push(Reverse((0, src as usize, 0)));

        while !min_heap.is_empty() {
            let Reverse((total_price, from, steps)) = min_heap.pop().unwrap();

            if steps > stops[from] || steps > k + 1 {
                continue;
            }
            if from as i32 == dst {
                return total_price;
            }

            stops[from] = steps;
            for &(to, price) in &graph[from] {
                min_heap.push(Reverse((total_price + price, to, steps + 1)));
            }
        }

        -1
    }
}
